src/main/java/com/qouteall/immersive_portals/Helper.java

```diff
@@ -658,4 +679,279 @@ public static Quaternion interpolateQuaternionNaive(
+
+        // If the portal is not closer, return the hit result we just got
+        if (hitResult.getPos().squaredDistanceTo(start) < intersection.squaredDistanceTo(start)) {
+            return new Pair<>(hitResult, portals);
+        }
+
+        // If the portal is closer, recurse
+
+        IERayTraceContext betterContext = (IERayTraceContext) context;
+
+        betterContext
+            .setStart(portal.transformPoint(intersection))
+            .setEnd(portal.transformPoint(end));
+
+        portals.add(portal);
+        EnvType env = world.isClient ? EnvType.CLIENT : EnvType.SERVER;
+        Pair<BlockHitResult, List<Portal>> recursion =
+            rayTrace(portal.getDestinationWorld(env), context, includeGlobalPortals, portals);
+
+        betterContext
+            .setStart(start)
+            .setEnd(end);
+
+        return recursion;
+    }
+
+    /**
+     * Ray traces for blocks or whatever the {@code context} dictates.
+     *
+     * @param world                The world to ray trace in.
+     * @param context              The ray tracing context to use. This context will be mutated as it goes but will be
+     *                             returned back to normal before a result is returned to you, so you can act like it
+     *                             hasn't been  mutated.
+     * @param includeGlobalPortals Whether or not to include global portals in the ray trace. If this is false, then the
+     *                             ray trace can pass right through them.
+     * @return The BlockHitResult and the list of portals that we've passed through to get there. This list can be used
+     * to transform looking directions or do whatever you want really.
+     * @author LoganDark
+     */
+    public static Pair<BlockHitResult, List<Portal>> rayTrace(
+        World world,
+        RayTraceContext context,
+        boolean includeGlobalPortals
+    ) {
+        return rayTrace(world, context, includeGlobalPortals, new ArrayList<>());
+    }
+
+    /**
+     * @param hitResult The HitResult to check.
+     * @return If the HitResult passed is either {@code null}, or of type {@link HitResult.Type#MISS}.
+     */
+    public static boolean hitResultIsMissedOrNull(HitResult hitResult) {
+        return hitResult == null || hitResult.getType() == HitResult.Type.MISS;
+    }
+
+    /**
+     * @param vec  The {@link Vec3d} to get the {@link Direction} of.
+     * @param axis The {@link Direction.Axis} of directions to exclude.
+     * @return The {@link Direction} of the passed {@code vec}, excluding directions of axis {@code axis}.
+     */
+    public static Direction getFacingExcludingAxis(Vec3d vec, Direction.Axis axis) {
+        Stream<Direction> directions = Arrays.stream(Direction.values()).filter(d -> !d.getAxis().equals(axis));
+        Direction samestDirection = Direction.NORTH;
+        double bestSameness = Double.MIN_VALUE;
+
+        for (Iterator<Direction> it = directions.iterator(); it.hasNext(); ) {
```

Comment:
Can be replaced by Stream.max
I like "beatSameness"

Reply:
Streams.max wants me to write a comparator which is dumb